### DISCUSSION TITLE
How to make a data set available in different modules AND user specific?

### DISCUSSION THREAD

**Discussion by zilch42** [zauberzeug/nicegui#2414] [Q&A] [answered]
### Question

Following on from #1844, I'm using a data class to hold all of my data objects for a given dataset and passing that to different page modules, but because that class object is created outside of the `pages`, it isn't user specific. When one user loads a dataset it changes the dataset that all uses see. How can I make it user specific?

The only solution I can think of is to create a Single Page App. The main reason I'm not doing that is because I don't like working in massive python files and I would rather break things down into modules. But it is just a preference I guess.

```python
# main.py
import data
import plot
from nicegui import ui

class refreshable_df_class():
    def __init__(self):
        self.df = None
    def load(self, df):
        self.df = df
        
data_model = refreshable_df_class()

@ui.page('/')
def main():
    ui.link("Dataset", "/dataset")
    ui.link("Plot", "/plot")

data.create(data_model)
plot.create(data_model)

ui.run()
```
```python
# data.py
import pandas as pd
from nicegui import ui


def create(data_model):
    @ui.refreshable
    def data_table_ui():
        if data_model.df is not None:
            ui.aggrid.from_pandas(df=data_model.df)
            
    def create_data(y: int):
        df = pd.DataFrame({'X': range(10), 'Y': [y]*10})
        data_model.load(df)
        data_table_ui.refresh()

    @ui.page('/dataset')
    def page_dataset():
        ui.button("Create Data 3's", on_click=lambda: create_data(3))
        ui.button("Create Data 7's", on_click=lambda: create_data(7))
        data_table_ui()
        ui.link("Home", "/")
```
```python
# plot.py
import pandas as pd
import plotly.graph_objects as go
from nicegui import ui
from typing import Union

@ui.refreshable
def plot_ui(data_model: Union[pd.DataFrame, None]):
    if data_model.df is not None:
        fig = go.Figure(go.Scatter(x=data_model.df.X, y=data_model.df.Y))
        fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
        ui.plotly(fig).classes('w-full h-40')

def create(data_model: Union[pd.DataFrame, None]):
    @ui.page('/plot')
    def pg_plot():
        plot_ui(data_model)
        ui.link("Home", "/")
```

**Comment by rodja** [comment, marked as answer]
You could use `app.storage.user`  to save all data user specific. And if your data is not easily storable in a dict you could build your own store and use `app.storage.browser['id']` as a unique key to find the right dataset.

**Reply by zilch42** [reply]
Hmmm that could work. I don't think saving the dataset to the user would be practical in my case because some of the datasets are hundreds of MB, maybe instead of loading all of the resources in the dataset once into the `data_model` and then passing that around, I just save the reference to the dataset in `app.storage.user` and then load the resources required on each page. It might result in more IO if they change pages regularly but the IO isn't too slow at the moment so that might be a better strategy. Thanks @rodja
